### PR TITLE
Move the handling of MPI_PROC_NULL out from MPID for RMA

### PR DESCRIPTION
**Pull request by huansong**
## Pull Request Description
_Updated_:
This PR moves all the current `MPID`-involved handling of `MPI_PROC_NULL` out from `MPID` layer up to the `MPI` layer for **RMA operations** (`MPID_[send|recv|...]` are used with `MPI_PROC_NULL` for collectives). The only exception is `MPI_Win_shared_query`.
*****
_Original_:
This PR cleans up the handling logic of `MPI_PROC_NULL` in ch4. The current ch4 has `MPI_PROC_NULL` checks all over the place and it often happens multiple times in the calling stack of one MPI operation. We may move all handling logic of `MPI_PROC_NULL` in ch4 to the `MPID_*` layer and gets rid of the redundant checks otherwise.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by huansong** [comment]
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
Why is this at the `MPID_` layer, instead of at the `MPI_` layer?

**Comment by huansong** [comment]
@pavanbalaji I think logically it can also be put at the `MPI_` layer, but it seems to me that some MPI operations need to access device-specific information when handling `MPI_PROC_NULL`. For example, `MPI_Win_shared_query` will access `MPIR_Win.basic_info_table` in ch3 but `MPIR_Win.dev.am.shared_table` in ch4. Similarly for `MPI_Win_lock`, `MPI_Win_unlock` etc.

**Comment by pavanbalaji** [comment]
Do `MPI_Win_lock` and `MPI_Win_unlock` need device-specific information?  I agree that `MPI_Win_shared_query` needs it, but maybe that's the only exception?

**Comment by hajimefu** [comment]
One thing I remember in a conversation with @minsii was that even for an RMA operation targeting `MPI_PROC_NULL`, error checks like "am I in a correct epoch to do this?" must happen according to the standard. And for now, epoch state is maintained at the device layer, not at MPI layer.

BTW @huansong it looks like this kind of check is missing in the current PR.

**Comment by pavanbalaji** [comment]
> One thing I remember in a conversation with @minsii was that even for an RMA operation targeting MPI_PROC_NULL, error checks like "am I in a correct epoch to do this?" must happen according to the standard. And for now, epoch state is maintained at the device layer, not at MPI layer.

That's not correct.  In fact, the standard doesn't mandate any error checking except for Fortran 90.

**Comment by huansong** [comment]
> Do `MPI_Win_lock` and `MPI_Win_unlock` need device-specific information? I agree that `MPI_Win_shared_query` needs it, but maybe that's the only exception?

Yes, `MPI_Win_lock|unlock` needs to update `MPIR_Win.lock_epoch_count` and `MPIR_Win.states.access_state` in ch3 but `MPIR_Win.dev.am.sync.lock.count` and `MPIR_Win.dev.am.sync.access_epoch_type` in ch4.

**Comment by pavanbalaji** [comment]
It should not *need to* update the epoch count.  That's not correct.  You are doing that as additional error checking.  I'd recommend dropping that and returning directly from the MPI layer.

**Comment by minsii** [comment]
@pavanbalaji Does standard define how the implementation should perform `error-checking` ? E.g., is it required that all `normal rank` cases should be error checked but the `PROC_NULL` case should not be checked ?

**Comment by huansong** [comment]
> One thing I remember in a conversation with @minsii was that even for an RMA operation targeting `MPI_PROC_NULL`, error checks like "am I in a correct epoch to do this?" must happen according to the standard. And for now, epoch state is maintained at the device layer, not at MPI layer.
> 
> BTW @huansong it looks like this kind of check is missing in the current PR.

Yes, the current PR does not do that. However, it seems to me that even the original code does not check the epoch for `MPI_PROC_NULL` (e.g., `MPIDI_OFI|UCX_do_put` checks that, but `MPIDI_put_unsafe` already returns for `MPI_PROC_NULL`).

**Comment by pavanbalaji** [comment]
> @pavanbalaji Does standard define how the implementation should perform error-checking ? E.g., is it required that all normal rank cases should be error checked but the PROC_NULL case should not be checked ?

It does not *require* anything with respect to error checking.  If you read my response to @hajimefu carefully, it said just that.  Specifically commenting on his statement that said:

`error checks ... must happen according to the standard`

There is no *must* in the standard.

**Comment by minsii** [comment]
I could not find a very detailed definition in the standard. If the intention is to let the implementation to decide how strict the error checking should be, then MPICH actually handles it very strictly. I.e., for `MPI_PROC_NULL` case, we still require the user to make correct epoch-open/closing calls.  For instance, the following program reports error with the default MPICH build:

```
win_create(...&win);
put(...MPI_PROC_NULL, ..., win);
```

**Comment by pavanbalaji** [comment]
> If the intention is to let the implementation to decide how strict the error checking should be, then MPICH actually handles it very strictly

Correct.  And in doing so, it is pushing the handling of `MPI_PROC_NULL` to the device.  My proposal is to not do that and return directly from the MPI layer.

**Comment by minsii** [comment]
@pavanbalaji I think it may need more changes in MPICH than just following the current rule (at least for RMA, the intention is to do error checking for `MPI_PROC_NULL`).

**Comment by pavanbalaji** [comment]
> @pavanbalaji I think it may need more changes in MPICH than just following the current rule (at least for RMA, the intention is to do error checking for MPI_PROC_NULL).

If it returns from the MPI layer, then the device-level code should not need any large changes because it's presumably not checking the rank before checking the epoch anyway.

**Comment by minsii** [comment]
The `MPID_*` layer code might be mess now, the original `MPID_*|NM_*|MPIDIG_*` code all check error for `MPI_PROC_NULL` (e.g., see `MPIDIG_do_put`).

**Comment by pavanbalaji** [comment]
Isn't this PR supposed to remove all those checks?

**Comment by huansong** [comment]
> Isn't this PR supposed to remove all those checks?

**Comment by huansong** [comment]
Deleted accidentally... I'all

**Comment by huansong** [comment]
@pavanbalaji I checked all MPI operations that need to handle `MPI_PROC_NULL`. It seems that `MPI_Win_shared_query` is the only exception. I will try to move all the others to `MPI_` layer and modify ch3 accordingly.

**Comment by minsii** [comment]
You are right...No way to check errors for `PROC_NULL` if we want to keep the `PROC_NULL` condition check only in one place. I am OK to move it to MPI layer and delete all `if(rank == PROC_NULL)` check in device layer.

**Comment by minsii** [comment]
Then we will also need to remove the unnecessary `PROC_NULL` check in RMA sync checking macros (e.g., `MPIDIG_RMA_OP_CHECK_SYNC`).

**Comment by huansong** [comment]
> Then we will also need to remove the unnecessary `PROC_NULL` check in RMA sync checking macros (e.g., `MPIDIG_RMA_OP_CHECK_SYNC`).

Yes this change is already included in the current PR.

**Comment by hajimefu** [comment]
> In fact, the standard doesn't mandate any error checking except for Fortran 90.

Hmm, I looked back the standard text and looks like @pavanbalaji is right. I might have misunderstood something in the past.

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/tcp

**Comment by huansong** [comment]
test:jenkins/ch3/tcp

**Comment by huansong** [comment]
Ready for review. @pavanbalaji @minsii

File: src/mpid/ch4/src/ch4_win.h [hunk 1]
```diff
@@ -210,7 +210,7 @@ MPL_STATIC_INLINE_PREFIX int MPID_Win_test(MPIR_Win * win, int *flag)
 #define FCNAME MPL_QUOTE(FUNCNAME)
 MPL_STATIC_INLINE_PREFIX int MPID_Win_lock(int lock_type, int rank, int assert, MPIR_Win * win)
 {
-    int mpi_errno;
+    int mpi_errno = MPI_SUCCESS;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_WIN_LOCK);
```
Comment (minsii) [inline src/mpid/ch4/src/ch4_win.h]: This change seems unrelated to this commit.

File: src/mpid/ch4/src/ch4_impl.h [hunk 2]
```diff
@@ -624,9 +618,8 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
     do {                                                                               \
         MPIDIG_EPOCH_CHECK_SYNC(win, mpi_errno, goto fn_fail);                     \
         MPIDIG_EPOCH_OP_REFENCE(win);                                              \
-        /* Check target sync status for any target_rank except PROC_NULL. */           \
-        if (target_rank != MPI_PROC_NULL)                                              \
-            MPIDIG_EPOCH_CHECK_TARGET_SYNC(win, target_rank, mpi_errno, goto fn_fail);  \
+        /* Check target sync status for any target_rank. */           \
+        MPIDIG_EPOCH_CHECK_TARGET_SYNC(win, target_rank, mpi_errno, goto fn_fail);  \
     } while (0);
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: I thought about this, but this change will not be valid alone if we had not done moving the `MPI_PROC_NULL` checks to `MPI_` for RMA routines. So it might be better to stay with the latter changes.
Other than this, I have addressed your other comments. The travis builds are passed (I think we may not need another round of Jenkins test). Could you please take another look?

[the remark this answers is not in the document]
Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: I do not think so. The reason we can move PROC_NULL check to `MPI_` layer is because we decided to give up error checking for PROC_NULL in RMA. So the change in error checking (and RMA code) should go first. Then you move all PROC_NULL checking to `MPI_` layer in the second commit.

[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: > The reason we can move PROC_NULL check to `MPI_` layer is because we decided to give up error checking for PROC_NULL in RMA.

I might be missing some point, but the original code here already means no error checking for `PROC_NULL`, right? So we are not "giving up" error checking for `PROC_NULL` here, but just removing an unnecessary check **after** we have done the moving action, so we know that an RMA op with `PROC_NULL` will not reach here. If the moving action had not been done, this change alone is invalid because `MPIDIG_EPOCH_CHECK_TARGET_SYNC` might not be able to handle `PROC_NULL`.

[the remark this answers is not in the document]
Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: In original code, we still check error in MPIDIG_EPOCH_CHECK_SYNC for PROC_NULL, but just do not check per-target status. Now, you skip MPIDIG_EPOCH_CHECK_SYNC if the target process is PROC_NULL.

[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: I see. I thought it was just about the code portion here. I have added a commit to skip all epoch checks for `PROC_NULL`. Does it look good?

[the remark this answers is not in the document]
Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: If `MPIDIG_RMA_OP_CHECK_SYNC` is always called after skip of `PROC_NULL` (e.g., in the third commit), you can skip the PROC_NULL check in the macro too. Can you please also change `MPIDIG_RMA_OP_CHECK_SYNC` in the third commit, and add comment saying that this macro assumes no `PROC_NULL` case calls it ?

[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: Done. I am re-running jenkins test to make sure the recent changes are good.

**Comment by huansong** [comment]
test:jenkins/ch4/most

**Comment by huansong** [comment]
@pavanbalaji Would you like to have a check on this PR?

**Comment by pavanbalaji** [comment]
Why is this PR only for RMA?

**Comment by huansong** [comment]
Discussed with @pavanbalaji offline. Now we will include P2P as well. Additional changes that need to be done for this: (1) Add `PROC_NULL` check in `MPIC_` P2P routines in order to guarantee correctness for collectives that call P2P with `PROC_NULL`. (2) Add `PROC_NULL` check in `MPI_Startall` or `MPID_Startall`.

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by huansong** [comment]
test:jenkins/ch3/tcp

**Comment by huansong** [comment]
test:jenkins/ch3/tcp

**Comment by minsii** [comment]
@huansong What is the status of this PR ? Can we merge it ?

**Comment by hzhou** [comment]
There are conflicts. Need rebase and re-run the tests.

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/pt2pt/sendrecv.c [hunk 3]
```diff
@@ -158,24 +158,39 @@ int MPI_Sendrecv(const void *sendbuf, int sendcount, MPI_Datatype sendtype,
...
-        MPIR_Request_free(rreq);
-        goto fn_fail;
-        /* --END ERROR HANDLING-- */
+    /* If source is MPI_PROC_NULL, create a completed request and return. */
+    if (unlikely(source == MPI_PROC_NULL)) {
+        rreq = MPIR_Request_create(MPIR_REQUEST_KIND__RECV);
+        MPIR_ERR_CHKANDSTMT(rreq == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+        MPIR_Object_set_ref(rreq, 1);
+        MPIR_cc_set(&rreq->cc, 0);
+        MPIR_Status_set_procnull(&rreq->status);
```
Comment (minsii) [inline src/mpi/pt2pt/sendrecv.c]: Use `MPIR_Request_create_complete` ?

File: src/mpid/ch4/src/ch4_impl.h [hunk 4]
```diff
@@ -429,12 +429,6 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIU_VALID_GROUP_RANK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIU_VALID_GROUP_RANK);
 
-    if (unlikely(rank == MPI_PROC_NULL)) {
-        /* Treat PROC_NULL as always valid */
-        ret = 1;
-        goto fn_exit;
-    }
-
```
[the remark this answers is not in the document]
Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: What do you think about this comment ?

**Comment by minsii** [comment]
@huansong Looks like the master branch has been updated. Can you please rebase it against master when doing next push ? I am learning the new review/approve process...

File: src/mpi/coll/helper_fns.c [hunk 5]
```diff
@@ -292,14 +314,32 @@ int MPIC_Sendrecv(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtype
...
-    if (mpi_errno)
-        MPIR_ERR_POP(mpi_errno);
+    /* If source is MPI_PROC_NULL, create a completed request and return. */
+    if (unlikely(source == MPI_PROC_NULL)) {
+        recv_req_ptr = MPIR_Request_create(MPIR_REQUEST_KIND__RECV);
+        MPIR_ERR_CHKANDSTMT(recv_req_ptr == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail,
+                            "**nomemreq");
+        MPIR_Object_set_ref(recv_req_ptr, 1);
+        MPIR_cc_set(&recv_req_ptr->cc, 0);
+        MPIR_Status_set_procnull(&recv_req_ptr->status);
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/mpi/coll/helper_fns.c]: There are some variables that are needed for creating empty RECV requests, and `MPIR_Request_create_complete` does not set those (I think mostly in `MPIR_Status_set_procnull`). This caused some issues in my tests earlier so I chose to create and set everything explicitly for RECV requests, just like the original code. However, I agree that `MPIR_Request_create_complete` would be much better in both code cleanness and performance. So I am now trying again with `MPIR_Request_create_complete` but also add `MPIR_Status_set_procnull` for RECV requests.

File: src/mpid/ch4/src/ch4_impl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: I think I initially restored this check after your comment. But later since we have included non-RMA functions in this PR, we can safely remove this check now, right?

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4_impl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: It is a utility function. For now it is used only in the RMA checking. But maybe in future it will be used in a function which has to handle PROC_NULL. How about we add a comment here saying that it assumes the calling routine never checks `PROC_NULL` ?

File: src/mpi/coll/helper_fns.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (minsii) [reply src/mpi/coll/helper_fns.c]: Yes, `MPIR_Request_create_complete` should not set status. So, in such empty receive, you will need both `MPIR_Request_create_complete` and `MPIR_Status_set_procnull`.

**Comment by huansong** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4_impl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (huansong) [reply src/mpid/ch4/src/ch4_impl.h]: I see. I have added the comment.

**Comment by hzhou** [comment]
@minsii You could merge this PR simply by click the Merge button. Enter signature like:
```
...
Approved-by: Min Si <email>
```

**Comment by minsii** [comment]
@hzhou Thanks. Merged.
